// astro/dist/vite-plugin-jsx/index.js
import babel from "@babel/core";
import * as eslexer from "es-module-lexer";
import esbuild from "esbuild";
import * as colors from "kleur/colors";
import path from "path";
import { error } from "../core/logger/core.js";
import { parseNpmName } from "../core/util.js";
import tagExportsPlugin from "./tag.js";
const JSX_EXTENSIONS = /* @__PURE__ */ new Set([".jsx", ".tsx", ".mdx"]);
const IMPORT_STATEMENTS = {
  react: "import React from 'react'",
  preact: "import { h } from 'preact'",
  "solid-js": "import 'solid-js'",
  astro: "import 'astro/jsx-runtime'"
};
const PREVENT_UNUSED_IMPORTS = ";;(React,Fragment,h);";
const IMPORT_KEYWORD_REGEX = /import/;
function getEsbuildLoader(fileExt) {
  if (fileExt === ".mdx")
    return "jsx";
  return fileExt.slice(1);
}
function collectJSXRenderers(renderers) {
  const renderersWithJSXSupport = renderers.filter((r) => r.jsxImportSource);
  return new Map(
    renderersWithJSXSupport.map((r) => [r.jsxImportSource, r])
  );
}
function detectImportSourceFromComments(code) {
  const multiline = code.match(/\/\*\*?[\S\s]*\*\//gm) || [];
  for (const comment of multiline) {
    const [_, lib] = comment.slice(0, -2).match(/@jsxImportSource\s*(\S+)/) || [];
    if (lib) {
      return lib.trim();
    }
  }
}
async function detectImportSourceFromImports(code, id, jsxRenderers) {
  const { code: jsCode } = await esbuild.transform(code + PREVENT_UNUSED_IMPORTS, {
    loader: getEsbuildLoader(path.extname(id)),
    jsx: "transform",
    jsxFactory: "h",
    jsxFragment: "Fragment",
    sourcefile: id,
    sourcemap: "inline"
  });
  const [imports] = eslexer.parse(jsCode);
  if (imports.length > 0) {
    for (let { n: spec } of imports) {
      const pkg = spec && parseNpmName(spec);
      if (!pkg)
        continue;
      if (jsxRenderers.has(pkg.name)) {
        return pkg.name;
      }
    }
  }
}
async function transformJSX({
  code,
  mode,
  id,
  ssr,
  renderer
}) {
  const { jsxTransformOptions } = renderer;
  const options = await jsxTransformOptions({ mode, ssr });
  const plugins = [...options.plugins || []];
  if (ssr) {
    plugins.push(tagExportsPlugin({ rendererName: renderer.name }));
  }
  const result = await babel.transformAsync(code, {
    presets: options.presets,
    plugins,
    cwd: process.cwd(),
    filename: id,
    ast: false,
    compact: false,
    sourceMaps: true,
    configFile: false,
    babelrc: false,
    inputSourceMap: options.inputSourceMap
  });
  if (!result)
    return null;
  if (renderer.name === "astro:jsx") {
    const { astro } = result.metadata;
    return {
      code: result.code || "",
      map: result.map,
      meta: {
        astro,
        vite: {
          lang: "ts"
        }
      }
    };
  }
  return {
    code: result.code || "",
    map: result.map
  };
}
function jsx({ settings, logging }) {
  let viteConfig;
  const jsxRenderers = /* @__PURE__ */ new Map();
  const jsxRenderersIntegrationOnly = /* @__PURE__ */ new Map();
  let astroJSXRenderer;
  let defaultJSXRendererEntry;
  return {
    name: "astro:jsx",
    enforce: "pre",
    async configResolved(resolvedConfig) {
      viteConfig = resolvedConfig;
      const possibleRenderers = collectJSXRenderers(settings.renderers);
      for (const [importSource, renderer] of possibleRenderers) {
        jsxRenderers.set(importSource, renderer);
        if (importSource === "astro") {
          astroJSXRenderer = renderer;
        } else {
          jsxRenderersIntegrationOnly.set(importSource, renderer);
        }
      }
      defaultJSXRendererEntry = [...jsxRenderersIntegrationOnly.entries()][0];
    },
    async transform(code, id, opts) {
      var _a;
      const ssr = Boolean(opts == null ? void 0 : opts.ssr);
      if (!JSX_EXTENSIONS.has(path.extname(id))) {
        return null;
      }
      const { mode } = viteConfig;
      if (id.includes(".mdx") || id.includes(".md")) {
        const { code: jsxCode2 } = await esbuild.transform(code, {
          loader: getEsbuildLoader(path.extname(id)),
          jsx: "preserve",
          sourcefile: id,
          sourcemap: "inline"
        });
        return transformJSX({
          code: jsxCode2,
          id,
          renderer: astroJSXRenderer,
          mode,
          ssr
        });
      }
      if (defaultJSXRendererEntry && jsxRenderersIntegrationOnly.size === 1) {
        const { code: jsxCode2 } = await esbuild.transform(code, {
          loader: getEsbuildLoader(path.extname(id)),
          jsx: "preserve",
          sourcefile: id,
          sourcemap: "inline"
        });
        return transformJSX({
          code: jsxCode2,
          id,
          renderer: defaultJSXRendererEntry[1],
          mode,
          ssr
        });
      }
      let importSource = detectImportSourceFromComments(code);
      if (!importSource && IMPORT_KEYWORD_REGEX.test(code)) {
        importSource = await detectImportSourceFromImports(code, id, jsxRenderers);
      }
      if (!importSource) {
        const compilerOptions = (_a = settings.tsConfig) == null ? void 0 : _a.compilerOptions;
        importSource = compilerOptions == null ? void 0 : compilerOptions.jsxImportSource;
      }
      if (!importSource && defaultJSXRendererEntry) {
        const [defaultRendererName] = defaultJSXRendererEntry;
        error(
          logging,
          "renderer",
          `${colors.yellow(id)}
Unable to resolve a renderer that handles this file! With more than one renderer enabled, you should include an import or use a pragma comment.
Add ${colors.cyan(
            IMPORT_STATEMENTS[defaultRendererName] || `import '${defaultRendererName}';`
          )} or ${colors.cyan(`/* jsxImportSource: ${defaultRendererName} */`)} to this file.
`
        );
        return null;
      } else if (!importSource) {
        error(
          logging,
          "renderer",
          `${colors.yellow(id)}
Unable to find a renderer for JSX. Do you have one configured in your Astro config? See this page to learn how:
https://docs.astro.build/en/core-concepts/framework-components/#installing-integrations
`
        );
        return null;
      }
      const selectedJsxRenderer = jsxRenderers.get(importSource);
      if (!selectedJsxRenderer) {
        error(
          logging,
          "renderer",
          `${colors.yellow(
            id
          )} No renderer installed for ${importSource}. Try adding \`@astrojs/${importSource}\` to your project.`
        );
        return null;
      }
      const { code: jsxCode } = await esbuild.transform(code, {
        loader: getEsbuildLoader(path.extname(id)),
        jsx: "preserve",
        sourcefile: id,
        sourcemap: "inline"
      });
      return await transformJSX({
        code: jsxCode,
        id,
        renderer: selectedJsxRenderer,
        mode,
        ssr
      });
    }
  };
}
export {
  jsx as default
};
