// astro/dist/core/errors.js
import eol from "eol";
import fs from "fs";
import { fileURLToPath } from "node:url";
import stripAnsi from "strip-ansi";
import { createLogger } from "vite";
import { codeFrame, createSafeError } from "./util.js";
var AstroErrorCodes = /* @__PURE__ */ ((AstroErrorCodes2) => {
  AstroErrorCodes2[AstroErrorCodes2["UnknownError"] = 1e3] = "UnknownError";
  AstroErrorCodes2[AstroErrorCodes2["ConfigError"] = 1001] = "ConfigError";
  AstroErrorCodes2[AstroErrorCodes2["UnknownCompilerError"] = 2e3] = "UnknownCompilerError";
  AstroErrorCodes2[AstroErrorCodes2["UnknownCompilerCSSError"] = 2001] = "UnknownCompilerCSSError";
  return AstroErrorCodes2;
})(AstroErrorCodes || {});
function cleanErrorStack(stack) {
  return stack.split(/\n/g).map((l) => l.replace(/\/@fs\//g, "/")).join("\n");
}
function fixViteErrorMessage(_err, server, filePath) {
  var _a, _b;
  const err = createSafeError(_err);
  try {
    server == null ? void 0 : server.ssrFixStacktrace(err);
  } catch {
  }
  if (err.message === "import.meta.glob() can only accept string literals.") {
    err.message = "Astro.glob() and import.meta.glob() can only accept string literals.";
  }
  if (filePath && /failed to load module for ssr:/.test(err.message)) {
    const importName = (_a = err.message.split("for ssr:").at(1)) == null ? void 0 : _a.trim();
    if (importName) {
      try {
        const content = fs.readFileSync(fileURLToPath(filePath)).toString();
        const lns = content.split("\n");
        const line = lns.findIndex((ln) => ln.includes(importName));
        if (line == -1)
          return err;
        const column = (_b = lns[line]) == null ? void 0 : _b.indexOf(importName);
        if (!err.id) {
          err.id = `${fileURLToPath(filePath)}:${line + 1}:${column + 1}`;
        }
      } catch {
      }
    }
  }
  return err;
}
const incompatiblePackages = {
  "react-spectrum": `@adobe/react-spectrum is not compatible with Vite's server-side rendering mode at the moment. You can still use React Spectrum from the client. Create an island React component and use the client:only directive. From there you can use React Spectrum.`
};
const incompatPackageExp = new RegExp(`(${Object.keys(incompatiblePackages).join("|")})`);
function createCustomViteLogger(logLevel) {
  const viteLogger = createLogger(logLevel);
  const logger = {
    ...viteLogger,
    error(msg, options) {
      if (incompatPackageExp.test(msg))
        return;
      return viteLogger.error(msg, options);
    }
  };
  return logger;
}
function generateHint(err, filePath) {
  var _a, _b;
  if (/Unknown file extension \"\.(jsx|vue|svelte|astro|css)\" for /.test(err.message)) {
    return "You likely need to add this package to `vite.ssr.noExternal` in your astro config file.";
  } else if (err.toString().startsWith("ReferenceError") && ((_b = ((_a = err.loc) == null ? void 0 : _a.file) ?? (filePath == null ? void 0 : filePath.pathname)) == null ? void 0 : _b.endsWith(".astro"))) {
    return "export statements in `.astro` files do not have access to local variable declarations, only imported values.";
  } else {
    const res = incompatPackageExp.exec(err.stack);
    if (res) {
      const key = res[0];
      return incompatiblePackages[key];
    }
  }
  return void 0;
}
function collectErrorMetadata(e, filePath) {
  var _a, _b;
  const err = e;
  if (e.stack) {
    e.stack = eol.lf(e.stack);
    const stackText = stripAnsi(e.stack);
    const possibleFilePath = err.pluginCode || err.id || stackText.split("\n").find((ln) => ln.includes("src") || ln.includes("node_modules"));
    const source = possibleFilePath == null ? void 0 : possibleFilePath.replace(/^[^(]+\(([^)]+).*$/, "$1").replace(/^\s+at\s+/, "");
    const [file, line, column] = (source == null ? void 0 : source.split(":")) ?? [];
    if (!err.loc && line && column) {
      err.loc = {
        file,
        line: Number.parseInt(line),
        column: Number.parseInt(column)
      };
    }
    if (!err.plugin) {
      err.plugin = ((_a = /withastro\/astro\/packages\/integrations\/([\w-]+)/gim.exec(stackText)) == null ? void 0 : _a.at(1)) || ((_b = /(@astrojs\/[\w-]+)\/(server|client|index)/gim.exec(stackText)) == null ? void 0 : _b.at(1)) || void 0;
    }
    err.stack = cleanErrorStack(e.stack);
  }
  if (e.name === "YAMLException") {
    err.loc = { file: e.id, line: e.mark.line, column: e.mark.column };
    err.message = e.reason;
  }
  if (!err.frame && err.loc) {
    try {
      const fileContents = fs.readFileSync(err.loc.file, "utf8");
      const frame = codeFrame(fileContents, err.loc);
      err.frame = frame;
    } catch {
    }
  }
  if (Array.isArray(e.errors)) {
    const { location, pluginName, text } = e.errors[0];
    if (location) {
      err.loc = { file: location.file, line: location.line, column: location.column };
      err.id = err.id || (location == null ? void 0 : location.file);
    }
    const possibleFilePath = err.pluginCode || err.id || (location == null ? void 0 : location.file);
    if (possibleFilePath && !err.frame) {
      try {
        const fileContents = fs.readFileSync(possibleFilePath, "utf8");
        err.frame = codeFrame(fileContents, err.loc);
      } catch {
      }
    }
    if (pluginName) {
      err.plugin = pluginName;
    }
    err.hint = generateHint(err, filePath);
    return err;
  }
  err.hint = generateHint(e, filePath);
  return err;
}
function getViteErrorPayload(err) {
  let plugin = err.plugin;
  if (!plugin && err.hint) {
    plugin = "astro";
  }
  const message = `${err.message}

${err.hint ?? ""}`;
  return {
    type: "error",
    err: {
      ...err,
      plugin,
      message: message.trim(),
      stack: err.stack
    }
  };
}
export {
  AstroErrorCodes,
  cleanErrorStack,
  collectErrorMetadata,
  createCustomViteLogger,
  fixViteErrorMessage,
  getViteErrorPayload
};
