// astro/dist/vite-plugin-astro-server/index.js
import mime from "mime";
import { Readable } from "stream";
import { getSetCookiesFromResponse } from "../core/cookies/index.js";
import { call as callEndpoint } from "../core/endpoint/dev/index.js";
import {
  collectErrorMetadata,
  fixViteErrorMessage,
  getViteErrorPayload
} from "../core/errors.js";
import { error, info, warn } from "../core/logger/core.js";
import * as msg from "../core/messages.js";
import { appendForwardSlash } from "../core/path.js";
import { getParamsAndProps, GetParamsAndPropsError } from "../core/render/core.js";
import { preload, ssr } from "../core/render/dev/index.js";
import { RouteCache } from "../core/render/route-cache.js";
import { createRequest } from "../core/request.js";
import { createRouteManifest, matchAllRoutes } from "../core/routing/index.js";
import { resolvePages } from "../core/util.js";
import notFoundTemplate, { subpathNotUsedTemplate } from "../template/4xx.js";
function writeHtmlResponse(res, statusCode, html) {
  res.writeHead(statusCode, {
    "Content-Type": "text/html; charset=utf-8",
    "Content-Length": Buffer.byteLength(html, "utf-8")
  });
  res.write(html);
  res.end();
}
async function writeWebResponse(res, webResponse) {
  const { status, headers, body } = webResponse;
  let _headers = {};
  if ("raw" in headers) {
    for (const [key, value] of Object.entries(headers.raw())) {
      res.setHeader(key, value);
    }
  } else {
    _headers = Object.fromEntries(headers.entries());
  }
  const setCookieHeaders = Array.from(getSetCookiesFromResponse(webResponse));
  if (setCookieHeaders.length) {
    res.setHeader("Set-Cookie", setCookieHeaders);
  }
  res.writeHead(status, _headers);
  if (body) {
    if (Symbol.for("astro.responseBody") in webResponse) {
      let stream = webResponse[Symbol.for("astro.responseBody")];
      for await (const chunk of stream) {
        res.write(chunk.toString());
      }
    } else if (body instanceof Readable) {
      body.pipe(res);
      return;
    } else if (typeof body === "string") {
      res.write(body);
    } else {
      const reader = body.getReader();
      while (true) {
        const { done, value } = await reader.read();
        if (done)
          break;
        if (value) {
          res.write(value);
        }
      }
    }
  }
  res.end();
}
async function writeSSRResult(webResponse, res) {
  return writeWebResponse(res, webResponse);
}
async function handle404Response(origin, settings, req, res) {
  const pathname = decodeURI(new URL(origin + req.url).pathname);
  const html = notFoundTemplate({
    statusCode: 404,
    title: "Not found",
    tabTitle: "404: Not Found",
    pathname
  });
  writeHtmlResponse(res, 404, html);
}
async function handle500Response(viteServer, origin, req, res, err) {
  res.on("close", () => setTimeout(() => viteServer.ws.send(getViteErrorPayload(err)), 200));
  if (res.headersSent) {
    res.write(`<script type="module" src="/@vite/client"><\/script>`);
    res.end();
  } else {
    writeHtmlResponse(
      res,
      500,
      `<title>${err.name}</title><script type="module" src="/@vite/client"><\/script>`
    );
  }
}
function getCustom404Route({ config }, manifest) {
  const relPages = resolvePages(config).href.replace(config.root.href, "");
  const pattern = new RegExp(`${appendForwardSlash(relPages)}404.(astro|md)`);
  return manifest.routes.find((r) => r.component.match(pattern));
}
function log404(logging, pathname) {
  info(logging, "serve", msg.req({ url: pathname, statusCode: 404 }));
}
function baseMiddleware(settings, logging) {
  const { config } = settings;
  const site = config.site ? new URL(config.base, config.site) : void 0;
  const devRoot = site ? site.pathname : "/";
  return function devBaseMiddleware(req, res, next) {
    var _a;
    const url = req.url;
    const pathname = decodeURI(new URL(url, "http://vitejs.dev").pathname);
    if (pathname.startsWith(devRoot)) {
      req.url = url.replace(devRoot, "/");
      return next();
    }
    if (pathname === "/" || pathname === "/index.html") {
      log404(logging, pathname);
      const html = subpathNotUsedTemplate(devRoot, pathname);
      return writeHtmlResponse(res, 404, html);
    }
    if ((_a = req.headers.accept) == null ? void 0 : _a.includes("text/html")) {
      log404(logging, pathname);
      const html = notFoundTemplate({
        statusCode: 404,
        title: "Not found",
        tabTitle: "404: Not Found",
        pathname
      });
      return writeHtmlResponse(res, 404, html);
    }
    next();
  };
}
async function matchRoute(pathname, routeCache, viteServer, logging, manifest, settings) {
  const matches = matchAllRoutes(pathname, manifest);
  for await (const maybeRoute of matches) {
    const filePath = new URL(`./${maybeRoute.component}`, settings.config.root);
    const preloadedComponent = await preload({ settings, filePath, viteServer });
    const [, mod] = preloadedComponent;
    const paramsAndPropsRes = await getParamsAndProps({
      mod,
      route: maybeRoute,
      routeCache,
      pathname,
      logging,
      ssr: settings.config.output === "server"
    });
    if (paramsAndPropsRes !== GetParamsAndPropsError.NoMatchingStaticPath) {
      return {
        route: maybeRoute,
        filePath,
        preloadedComponent,
        mod
      };
    }
  }
  if (matches.length) {
    warn(
      logging,
      "getStaticPaths",
      `Route pattern matched, but no matching static path found. (${pathname})`
    );
  }
  log404(logging, pathname);
  const custom404 = getCustom404Route(settings, manifest);
  if (custom404) {
    const filePath = new URL(`./${custom404.component}`, settings.config.root);
    const preloadedComponent = await preload({ settings, filePath, viteServer });
    const [, mod] = preloadedComponent;
    return {
      route: custom404,
      filePath,
      preloadedComponent,
      mod
    };
  }
  return void 0;
}
async function handleRequest(routeCache, viteServer, logging, manifest, settings, req, res) {
  var _a;
  const { config } = settings;
  const origin = `${viteServer.config.server.https ? "https" : "http"}://${req.headers.host}`;
  const buildingToSSR = config.output === "server";
  const url = new URL(origin + ((_a = req.url) == null ? void 0 : _a.replace(/(index)?\.html$/, "")));
  const pathname = decodeURI(url.pathname);
  url.pathname = config.base.substring(0, config.base.length - 1) + url.pathname;
  if (!buildingToSSR && pathname !== "/_image") {
    const allSearchParams = Array.from(url.searchParams);
    for (const [key] of allSearchParams) {
      url.searchParams.delete(key);
    }
  }
  let body = void 0;
  if (!(req.method === "GET" || req.method === "HEAD")) {
    let bytes = [];
    await new Promise((resolve) => {
      req.on("data", (part) => {
        bytes.push(part);
      });
      req.on("end", resolve);
    });
    body = Buffer.concat(bytes);
  }
  let filePath;
  try {
    const matchedRoute = await matchRoute(
      pathname,
      routeCache,
      viteServer,
      logging,
      manifest,
      settings
    );
    filePath = matchedRoute == null ? void 0 : matchedRoute.filePath;
    return await handleRoute(
      matchedRoute,
      url,
      pathname,
      body,
      origin,
      routeCache,
      viteServer,
      manifest,
      logging,
      settings,
      req,
      res
    );
  } catch (_err) {
    const err = fixViteErrorMessage(_err, viteServer, filePath);
    const errorWithMetadata = collectErrorMetadata(err);
    error(logging, null, msg.formatErrorMessage(errorWithMetadata));
    handle500Response(viteServer, origin, req, res, errorWithMetadata);
  }
}
async function handleRoute(matchedRoute, url, pathname, body, origin, routeCache, viteServer, manifest, logging, settings, req, res) {
  if (!matchedRoute) {
    return handle404Response(origin, settings, req, res);
  }
  const { config } = settings;
  const filePath = matchedRoute.filePath;
  const { route, preloadedComponent, mod } = matchedRoute;
  const buildingToSSR = config.output === "server";
  const request = createRequest({
    url,
    headers: buildingToSSR ? req.headers : new Headers(),
    method: req.method,
    body,
    logging,
    ssr: buildingToSSR,
    clientAddress: buildingToSSR ? req.socket.remoteAddress : void 0
  });
  const paramsAndPropsRes = await getParamsAndProps({
    mod,
    route,
    routeCache,
    pathname,
    logging,
    ssr: config.output === "server"
  });
  const options = {
    settings,
    filePath,
    logging,
    mode: "development",
    origin,
    pathname,
    route,
    routeCache,
    viteServer,
    request
  };
  if (route.type === "endpoint") {
    const result = await callEndpoint(options);
    if (result.type === "response") {
      if (result.response.headers.get("X-Astro-Response") === "Not-Found") {
        const fourOhFourRoute = await matchRoute(
          "/404",
          routeCache,
          viteServer,
          logging,
          manifest,
          settings
        );
        return handleRoute(
          fourOhFourRoute,
          new URL("/404", url),
          "/404",
          body,
          origin,
          routeCache,
          viteServer,
          manifest,
          logging,
          settings,
          req,
          res
        );
      }
      await writeWebResponse(res, result.response);
    } else {
      let contentType = "text/plain";
      const filepath = route.pathname || route.segments.map((segment) => segment.map((p) => p.content).join("")).join("/");
      const computedMimeType = mime.getType(filepath);
      if (computedMimeType) {
        contentType = computedMimeType;
      }
      res.writeHead(200, { "Content-Type": `${contentType};charset=utf-8` });
      res.end(result.body);
    }
  } else {
    const result = await ssr(preloadedComponent, options);
    return await writeSSRResult(result, res);
  }
}
function createPlugin({ settings, logging }) {
  return {
    name: "astro:server",
    configureServer(viteServer) {
      let routeCache = new RouteCache(logging, "development");
      let manifest = createRouteManifest({ settings }, logging);
      function rebuildManifest(needsManifestRebuild, file) {
        routeCache.clearAll();
        if (needsManifestRebuild) {
          manifest = createRouteManifest({ settings }, logging);
        }
      }
      viteServer.watcher.on("add", rebuildManifest.bind(null, true));
      viteServer.watcher.on("unlink", rebuildManifest.bind(null, true));
      viteServer.watcher.on("change", rebuildManifest.bind(null, false));
      return () => {
        if (settings.config.base !== "/") {
          viteServer.middlewares.stack.unshift({
            route: "",
            handle: baseMiddleware(settings, logging)
          });
        }
        viteServer.middlewares.use(async (req, res) => {
          if (!req.url || !req.method) {
            throw new Error("Incomplete request");
          }
          handleRequest(routeCache, viteServer, logging, manifest, settings, req, res);
        });
      };
    },
    transform(code, id, opts = {}) {
      if (opts.ssr)
        return;
      if (!id.includes("vite/dist/client/client.mjs"))
        return;
      return code.replace(/\.tip \{[^}]*\}/gm, ".tip {\n  display: none;\n}").replace(/\[vite\]/g, "[astro]");
    }
  };
}
export {
  baseMiddleware,
  createPlugin as default
};
