// astro/dist/vite-plugin-scripts/page-ssr.js
import { PAGE_SSR_SCRIPT_ID } from "./index.js";
import ancestor from "common-ancestor-path";
import MagicString from "magic-string";
import { isPage } from "../core/util.js";
function astroScriptsPostPlugin({
  settings
}) {
  function normalizeFilename(filename) {
    if (filename.startsWith("/@fs")) {
      filename = filename.slice("/@fs".length);
    } else if (filename.startsWith("/") && !ancestor(filename, settings.config.root.pathname)) {
      filename = new URL("." + filename, settings.config.root).pathname;
    }
    return filename;
  }
  return {
    name: "astro:scripts:page-ssr",
    enforce: "post",
    transform(code, id, options) {
      if (!(options == null ? void 0 : options.ssr))
        return;
      const hasInjectedScript = settings.scripts.some((s2) => s2.stage === "page-ssr");
      if (!hasInjectedScript)
        return;
      const filename = normalizeFilename(id);
      let fileURL;
      try {
        fileURL = new URL(`file://${filename}`);
      } catch (e) {
        return;
      }
      const fileIsPage = isPage(fileURL, settings);
      if (!fileIsPage)
        return;
      const s = new MagicString(code, { filename });
      s.prepend(`import '${PAGE_SSR_SCRIPT_ID}';
`);
      return {
        code: s.toString(),
        map: s.generateMap()
      };
    }
  };
}
export {
  astroScriptsPostPlugin as default
};
